### PR TITLE
ir: When something has a definition, return unresolved type references until we parse it.

### LINKED ISSUE #888: Global enum in the wrong namespace
Hello,

While attempting to generate rust bindings for [Halide](http://halide-lang.org/), I think that I encountered a bug. It seems like when the declaration and definition of a type occurs in two separate namespace 'wrappers' and the definition uses an `enum` from the global scope some bogus rust code is generated.

### Input C/C++ Header

```C++
namespace Halide {
struct Type;
}
typedef enum {} a;
namespace Halide {
struct Type {
  static a b;
};
}
```

### Bindgen Invocation

```Rust
extern crate bindgen;

use std::env;
use std::path::PathBuf;

fn main()
{
    let bindings = bindgen::Builder::default()
        .header("test.hpp")
        .clang_args(["-x", "c++", "-std=c++11"].iter())
        .enable_cxx_namespaces()
        .whitelisted_type("Halide::Type")
        .generate()
        .expect("Unable to generate rust bindings");

    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings
        .write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

### Actual Results

```
error[E0412]: cannot find type `a` in module `root::Halide`
  --> /work/git/rust-halide/hello-cpp-rs/target/debug/build/hello-cpp-a0439f5fa3eaba26/out/bindings.rs:17:50
   |
17 |             pub static mut Type_b: root::Halide::a;
   |                                                  ^ not found in `root::Halide`
   |
help: possible candidate is found in another module, you can import it into scope
   |
7  |     use root::a;
   |
```

The generated bindings end up like this:

```rust
#[allow(non_snake_case, non_camel_case_types, non_upper_case_globals)]
pub mod root {
    #[allow(unused_imports)]
    use self::super::root;
    pub mod Halide {
        #[allow(unused_imports)]
        use self::super::super::root;
        #[repr(C)]
        #[derive(Debug, Copy)]
        pub struct Type {
            pub _address: u8,
        }
        extern "C" {
            #[link_name = "_ZN6Halide4Type1bE"]
            pub static mut Type_b: root::Halide::a;
        }
        #[test]
        fn bindgen_test_layout_Type() {
            assert_eq!(::std::mem::size_of::<Type>() , 1usize , concat ! (
                       "Size of: " , stringify ! ( Type ) ));
            assert_eq! (::std::mem::align_of::<Type>() , 1usize , concat ! (
                        "Alignment of " , stringify ! ( Type ) ));
        }
        impl Clone for Type {
            fn clone(&self) -> Self { *self }
        }
    }
    #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
    pub enum a { }
}
```

and/or

### Expected Results

I expected the member variable `b` to be of the type `root::a` rather than `root::Halide::a` since it is technically defined in the global namespace rather than inside the `Halide` namespace.

### PR DESCRIPTION
**Pull request by emilio**
This fixes #888

### CODE REVIEW

File: src/ir/item.rs [hunk 1]
```diff
@@ -1194,7 +1194,18 @@ impl ClangItemParser for Item {
         // Types are sort of special, so to avoid parsing template classes
         // twice, handle them separately.
         {
-            let applicable_cursor = cursor.definition().unwrap_or(cursor);
+            let definition = cursor.definition();
+            let applicable_cursor = definition.unwrap_or(cursor);
+
+            if definition.is_some() && definition != Some(cursor) {
```
Comment (fitzgen) [inline src/ir/item.rs]: Care to plop the commit message as a comment right here? I think that would help future readers of this code.

File: tests/expectations/tests/class_nested.rs [hunk 2]
```diff
@@ -73,6 +53,26 @@ fn bindgen_test_layout_A() {
 impl Clone for A {
     fn clone(&self) -> Self { *self }
 }
+#[repr(C)]
+#[derive(Debug, Default, Copy, Hash, PartialEq, Eq)]
+pub struct C {
```
Comment (fitzgen) [inline tests/expectations/tests/class_nested.rs]: This should still be `A_C` though, shouldn't it? I don't think this change is desirable.

```c++
class A {
public:
    int member_a;
    class B {
        int member_b;
    };

    class C;

    template<typename T>
    class D {
      T foo;
    };
};

class A::C {
  int baz;
};
```

Comment (emilio) [inline tests/expectations/tests/class_nested.rs]: Good catch here, thanks! I'll take a look tomorrow when I have the time.

**Comment by emilio** [comment]
I fixed that bit, can you re-review @fitzgen?

**Pull request merged**